### PR TITLE
DHIS2-4712 + DHIS2-5895 - support program indicator + relationship type in /enrollment 

### PR DESCRIPTION
**Pull request by luciano-fiandesio**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getAggregateClauseForProgramIndicator( ProgramIndicator pi, AnalyticsType outerSqlEntity,
+        Date earliestStartDate, Date latestDate )
+    {
+        return _getAggregateClauseForProgramIndicator( pi, null, outerSqlEntity, earliestStartDate, latestDate );
```
Comment (Bekkalizer) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java]: Is the underscore a new convention I am not familiar with, or should the method be given another name?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        AnalyticsType outerSqlEntity, Date earliestStartDate, Date latestDate )
+    {
+        // Define aggregation function (avg, sum, ...) //
+        String function = TextUtils.emptyIfEqual( pi.getAggregationTypeFallback().getValue(),
+            AggregationType.CUSTOM.getValue() );
+
+        // Get sql construct from Program indicator expression //
+        String aggregateSql = getPrgIndSql( pi.getExpression(), pi, earliestStartDate, latestDate );
+
+        aggregateSql += ")";
```
Comment (Bekkalizer) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java]: Where is the corresponding "(" coming from here?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java (hunk 1 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java]: Ops, good catch, I was supposed to rename it.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java (hunk 2 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/programIndicator/DefaultProgramIndicatorSubqueryBuilder.java]: Added a comment. This parenthesis closes the initial function parenthesis:

`avg( PI query )` <-- this parenthesis

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 3]
```diff
@@ -174,18 +182,147 @@ public void verifyWithProgramStageAndNumericDataElementAndFilter2() {
...
+        when( programIndicatorService.getAnalyticsSql( "", programIndicatorA, getDate( 2000, 1, 1 ),
+            getDate( 2017, 4, 8 ), "subax" ) ).thenReturn( piSubquery );
+
+        subject.getEnrollments( params.build(), new ListGrid(), 100 );
+
+        verify( jdbcTemplate ).queryForRowSet( sql.capture() );
+
+        String expected = "ax.\"monthly\",ax.\"ou\",(SELECT avg (" + piSubquery + ") FROM analytics_event_"
+            + programB.getUid().toLowerCase() + " as subax WHERE  "
+            + "ax.tei in (select tei.uid from trackedentityinstance tei "
```
Comment (Bekkalizer) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]: Was here expecting the join to link ax.tei to the correct subax row through the relationship.
Link to cell in K-map: https://docs.google.com/spreadsheets/d/18u4ZPB_11CfAlxtEz3G85nwmI82hgncHHgQcmVL_B6w/edit#gid=0&range=D3

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipEntity.java [hunk 4]
```diff
@@ -45,4 +49,19 @@ public String getName()
...
     }
+
+    static
+    {
+        Map<String, RelationshipEntity> map = new ConcurrentHashMap<>();
+        for ( RelationshipEntity instance : RelationshipEntity.values() )
+        {
+            map.put( instance.getName(), instance );
+        }
+        ENUM_MAP = Collections.unmodifiableMap( map );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipEntity.java]: I think you can use this method for ENUM_MAP and avoid the entire `static` block:

https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Maps.html#uniqueIndex-java.lang.Iterable-com.google.common.base.Function-

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+import static org.hisp.dhis.common.DimensionalObjectUtils.getDimensionalItemIds;
+import static org.hisp.dhis.common.IdentifiableObjectUtils.getLocalPeriodIdentifiers;
+import static org.hisp.dhis.common.IdentifiableObjectUtils.getUids;
+import static org.hisp.dhis.organisationunit.OrganisationUnit.getParentGraphMap;
+import static org.hisp.dhis.organisationunit.OrganisationUnit.getParentNameGraphMap;
+
+/**
+ * @author Luciano Fiandesio
+ */
+public abstract class AbstractAnalyticsService {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java]: New line of {

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipEntity.java (hunk 4 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipEntity.java]: Fixed using Stream.of instead of Guava

**Pull request merged**